### PR TITLE
Add new lint: `manual_highest_one`

### PR DESCRIPTION
**Pull request by qdot3**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/17472)*

implementation of https://github.com/rust-lang/rust-clippy/issues/16985#issue-4414925038

checklist

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally (except for `cargo test --test dogfood` because `tikv_jemalloc_sys` is not resolved)
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

changelog: [`manual_highest_one`]: add new lint to check manual implementation of `highest_one()`

### CODE REVIEW

**Comment by qdot3** [comment]
Sorry, I confused with `cargo test` and `uitest`.

File: custom.toml
Comment (Gri-ffin) [inline custom.toml:1]: Was this commited by mistake?

File: clippy_lints/src/manual_highest_one.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                "try",
+                format!("{recv}.highest_one().unwrap_or({nonzero_arm})"),
+                app,
+            );
+        }
+    }
+}
+
+/// Returns `x` of `BITS - 1 - x.leading_zeros()`-like expressions.
+fn extract_recv_from_highest_one_equiv<'a>(cx: &LateContext<'a>, expr: &'a Expr<'a>) -> Option<&'a Expr<'a>> {
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: The convention is `'cx` or `'tcx`.

File: clippy_lints/src/manual_highest_one.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            {
+                return Some(recv);
+            }
+        },
+        _ => (),
+    }
+    None
+}
+
+fn extract_recv_of_lz<'a>(expr: &'a Expr<'a>) -> Option<&'a Expr<'a>> {
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: Ditto.

File: clippy_lints/src/manual_highest_one.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    if let ExprKind::Lit(lit) = expr.kind
+        && let LitKind::Int(Pu128(lit), _) = lit.node
+    {
+        Some(lit)
+    } else {
+        None
+    }
+}
+
+fn check_lit_one_and_extract_recv_of_lz<'a>(expr1: &'a Expr<'a>, expr2: &'a Expr<'a>) -> Option<&'a Expr<'a>> {
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: Same here.

File: clippy_lints/src/manual_highest_one.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+fn extract_lit(expr: &Expr<'_>) -> Option<u128> {
+    if let ExprKind::Lit(lit) = expr.kind
+        && let LitKind::Int(Pu128(lit), _) = lit.node
+    {
+        Some(lit)
+    } else {
+        None
+    }
+}
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: Why not use `integer_const`?

**Review by Gri-ffin — COMMENTED** [review]
Also can you please add
```rust
let _ = u32::BITS - u.leading_zeros() - 1; //~ manual_highest_one
let _ = u32::BITS - (u.leading_zeros() + 1); //~ manual_highest_one
```
to the test suite.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..b0bea4495f75fc0f20db324e9f6bbd564dc1f842)*

File: custom.toml
Comment (qdot3) [reply custom.toml:1]: Yes, this file was created by mistake.

**Comment by qdot3** [comment]
I added some test cases and found that `u32::BITS - nz.leading_zeros()` incorrectly triggers the `mismatched_bit_width_type` lint, so I'm ignoring it for now.

File: clippy_lints/src/manual_highest_one.rs (hunk 2 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 3 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 4 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 1 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: tests/ui/manual_highest_one.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+#![warn(clippy::manual_highest_one)]
+// FIXME: remove this after this lint is fixed
+#![allow(clippy::mismatched_bit_width_type)]
```
Comment (Gri-ffin) [inline tests/ui/manual_highest_one.rs]: You can test `u32::BITS - nz.leading_zeros();` and similar cases, but please use `expect` instead of `allow` (when possible).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: clippy_lints/src/manual_highest_one.rs [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    cx: &LateContext<'_>,
+    expr1: &'tcx Expr<'tcx>,
+    expr2: &'tcx Expr<'tcx>,
+) -> Option<&'tcx Expr<'tcx>> {
+    if let Some(recv) = extract_recv_of_lz(expr1)
+        && Some(1) == integer_const(cx, expr2, expr2.span.ctxt())
+    {
+        Some(recv)
+    } else if let Some(recv) = extract_recv_of_lz(expr2)
+        && Some(1) == integer_const(cx, expr1, expr1.span.ctxt())
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: You should compare the context with the parent's.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: clippy_lints/src/manual_highest_one.rs [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    /// let i = 31 - x.leading_zeros();
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x: u32 = 5;
+    /// let i = x.highest_one().unwrap();
+    /// ```
+    #[clippy::version = "1.99.0"]
+    pub MANUAL_HIGHEST_ONE,
+    nursery,
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: It's best to choose a category other than nursery.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: clippy_lints/src/manual_highest_one.rs [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+                    expr: Some(false_arm), ..
+                },
+                ..,
+            ) = else_block.kind
+            && let ExprKind::Binary(bin_op, cond_lhs, cond_rhs) = cond.kind
+        {
+            let check_cond = |lhs, rhs| {
+                if Some(0) == integer_const(cx, lhs, lhs.span.ctxt()) {
+                    Some(rhs)
+                } else if Some(0) == integer_const(cx, rhs, rhs.span.ctxt()) {
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: same here, you should pass the parent's context, you can also use `is_zero_integer_const` from `clippy_utils`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: clippy_lints/src/manual_highest_one.rs [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+) -> Option<&'tcx Expr<'tcx>> {
+    match expr.kind {
+        // BITS - 1 - x.leading_zeros()
+        // BITS - 1 - nonzero.leading_zeros()
+        ExprKind::Binary(bin_op, lhs, rhs) if bin_op.node == BinOpKind::Sub => {
+            // literal - x.leading_zeros()
+            if let Some(recv) = extract_recv_of_lz(rhs)
+                    && let Some(lit) = integer_const(cx, lhs, lhs.span.ctxt())
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: Same here.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: clippy_lints/src/manual_highest_one.rs [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+    pub MANUAL_HIGHEST_ONE,
+    nursery,
+    "manually reimplementing `highest_one`"
+}
+
+impl_lint_pass!(ManualHighestOne => [MANUAL_HIGHEST_ONE]);
+
+pub struct ManualHighestOne {
+    msrv: Msrv,
+    already_linted: FxHashSet<HirId>,
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: The current approach has lint duplications which is...not ideal. Why not match the expression, if so try to walk upward, if its part of an `if` pattern, we lint the `if`, if not we lint the expression, instead of checking the contents of the `if` expression then later re-checking if the expression is not inside an if pattern?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: tests/ui/manual_highest_one.rs (hunk 5 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs]: done

File: clippy_lints/src/manual_highest_one.rs [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        if !self.msrv.meets(cx, msrvs::HIGHEST_ONE) || expr.span.from_expansion() || !mid_ty.is_integral() {
+            return;
+        }
+
+        if let Some(recv) = extract_recv_from_highest_one_equiv(cx, expr) {
+            if self.already_linted.contains(&expr.hir_id()) {
+                return;
+            }
+
+            let mut app = Applicability::MaybeIncorrect;
```
Comment (CommanderStorm) [inline clippy_lints/src/manual_highest_one.rs]: just curious: what makes this is maybe incoorect? 🤔

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/5a78cd3ba2f6984d69976f183e12c8edfc0c29e9..8d6990755d926512d640a885904af9158d69041d)*

File: clippy_lints/src/manual_highest_one.rs (hunk 6 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 7 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: I choose complexity.

File: clippy_lints/src/manual_highest_one.rs (hunk 8 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 9 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 10 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 11 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: I assumed `offset + u32::BITS - u.trailing_zeros() - 1`, but that was overthinking. Applicability is now switched to `MachineApplicable`.

File: tests/ui/manual_highest_one.rs [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+            may_change_state();
+            31 - u.leading_zeros() //~ manual_highest_one
+        },
+    };
+
+    // --- Macro ---
+    let _ = identity!(31) - u.leading_zeros(); //~ manual_highest_one
+    let _ = 31 - identity!(u).leading_zeros(); //~ manual_highest_one
+    let _ = 31 - double!(u).leading_zeros(); //~ manual_highest_one
+    let _ = thirty_one!() - u.leading_zeros();
```
Comment (Gri-ffin) [inline tests/ui/manual_highest_one.rs:137]: We should be more conservative about macro expanded subexpressions, especially if the lint is aiming to be `MachineApplicable`. So we should also skip receivers or const operands coming from expansion too.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/9117fa75789b37e502b24ac47015e778d80cdab1..ced8d1c49665d794af1a1d829ffb66bfbf028262)*

File: clippy_lints/src/manual_highest_one.rs [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+                        let var = if is_zero_integer_const(cx, lhs, cond.span.ctxt())
+                            && [Rel::Eq, Rel::Ne, Rel::Lt].contains(&rel)
+                        {
+                            rhs
+                        } else if is_zero_integer_const(cx, rhs, cond.span.ctxt()) && [Rel::Eq, Rel::Ne].contains(&rel)
+                        {
+                            lhs
+                        } else {
+                            break;
+                        };
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: This needs a signdness check, I tested with:
```rust
let x = -1_i32;
let _ = if x > 0 {
    31 - x.leading_zeros()
} else {
    0
};
```
but i got `let new = x.highest_one().unwrap_or_else(|| 0);`. These are not equivalent, treating `x > 0` as equivalent `x != 0` is only true for unsigned integers.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/9117fa75789b37e502b24ac47015e778d80cdab1..ced8d1c49665d794af1a1d829ffb66bfbf028262)*

File: clippy_lints/src/manual_highest_one.rs (hunk 7 above)
Comment (Gri-ffin) [reply clippy_lints/src/manual_highest_one.rs]: Its best to change it here.

File: clippy_lints/src/manual_highest_one.rs [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+                    && arms[0].guard.is_none()
+                    && let PatKind::Expr(pat_expr) = arms[0].pat.kind
+                    && !pat_expr.span.from_expansion()
+                    && ConstEvalCtxt::new(cx).eval_pat_expr(pat_expr) == Some(Constant::Int(0))
+                    // `_` pattern
+                    && arms[1].guard.is_none()
+                    && let PatKind::Wild = arms[1].pat.kind
+                    // && arms[1].body.hir_id == h1_expr.hir_id
+                    // same item
+                    && matches!((key.res_local_id(), recv.res_local_id()), (Some(l), Some(r)) if l == r) =>
```
Comment (Gri-ffin) [inline clippy_lints/src/manual_highest_one.rs]: This code now would apply a wrong suggestion
```rust
let _ = match u {
    0 => 31 - u.leading_zeros(),
    _ => 123,
};
```
You aren't verifying that `h1_expr` belongs to the wildcard arm so I get: `let _ = u.highest_one().unwrap_or_else(|| 31 - u.leading_zeros());` You can use the arm id for the check `&& arms[1].hir_id == hir_id` instead of the comment. 

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/9117fa75789b37e502b24ac47015e778d80cdab1..ced8d1c49665d794af1a1d829ffb66bfbf028262)*

File: tests/ui/manual_highest_one.rs [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        let _ = u32::BITS - nz.leading_zeros();
+        let _ = u32::BITS - nz.leading_zeros() - 1; //~ manual_highest_one
+    }
+    let _ = u32::BITS - (1 + nz.leading_zeros()); //~ manual_highest_one
+    let _ = u32::BITS - (nz.leading_zeros() + 1); //~ manual_highest_one
+    // False negative
+    let _ = nz.bit_width().get() - 1;
+    // False positive: Currently, `integer_const` does not support `NonZero`
+    // and related aliases. It is better to enhance `integer_const`.
+    let _ = NonZeroU32::BITS - 1 - nz.leading_zeros();
```
Comment (Gri-ffin) [inline tests/ui/manual_highest_one.rs]: This should be handled, Instead of relying only on `integer_const`, maybe we can have a small helper that first tries integer_const, and if that fails, recognizes T::BITS directly and gets the bit width from the type:
```rust
fn integer_const_or_bits(cx: &LateContext<'_>, expr: &Expr<'_>, ctxt: SyntaxContext) -> Option<u128> {
    integer_const(cx, expr, ctxt).or_else(|| {
        if let ExprKind::Path(QPath::TypeRelative(hir_ty, segment)) = expr.kind
            && segment.ident.name == sym::BITS
        {
            bit_width(cx, cx.typeck_results().node_type(hir_ty.hir_id))
        } else {
            None
        }
    })
}
```
Then we can use it in the `bit_width` checks.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/9117fa75789b37e502b24ac47015e778d80cdab1..ced8d1c49665d794af1a1d829ffb66bfbf028262)*

File: clippy_lints/src/manual_highest_one.rs [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+impl ManualHighestOne {
+    pub fn new(conf: &Conf) -> Self {
+        Self { msrv: conf.msrv.into() }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for ManualHighestOne {
+    #[expect(clippy::too_many_lines)]
```
Comment (CommanderStorm) [inline clippy_lints/src/manual_highest_one.rs]: Can you break some part of this function out?
It is a bit hard to keep all of this in my minds context, which makes bugs much more likely.

File: tests/ui/manual_highest_one.rs [hunk 17]
```diff
@@ -0,0 +1,170 @@
+#![warn(clippy::manual_highest_one)]
+
+use std::num::NonZeroU32;
+
+fn do_something<T>(x: T) {
+    todo!()
+}
```
Comment (CommanderStorm) [inline tests/ui/manual_highest_one.rs]: can you use `black_box` here instead?
Not sure if the one below can also be `black_box`-ed instead

File: tests/ui/manual_highest_one.rs [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+    let _ = u.bit_width() - 1; //~ manual_highest_one
+    // False negative
+    let _ = u32::BITS.checked_sub(u.leading_zeros() + 1);
+
+    // --- Nonzero type ---
+    let _ = 31 - nz.leading_zeros(); //~ manual_highest_one
+    let _ = u32::BITS - 1 - nz.leading_zeros(); //~ manual_highest_one
+    {
+        #![expect(clippy::mismatched_bit_width_type)]
+        // this trigger `mismatched_bit_width_type`
```
Comment (CommanderStorm) [inline tests/ui/manual_highest_one.rs]: lets do it this way instead.

Also, why is this in a scope? Is this relevant for the lint?

```suggestion
        #![expect(clippy::mismatched_bit_width_type, reason = "...")]
```

File: tests/ui/manual_highest_one.rs [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+    // False negative
+    #[expect(clippy::blocks_in_conditions)]
+    let _ = if {
+        do_something(u);
+        u == 0
+    } {
+        todo!()
+    } else {
+        31 - u.leading_zeros() //~ manual_highest_one
+    };
```
Comment (CommanderStorm) [inline tests/ui/manual_highest_one.rs]: 🤔  What do you mean with `False negative`? It does seem to lint..

File: tests/ui/manual_highest_one.rs [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+    // False negatives
+    let _ = if u == 0 {
+        let v = u;
+        31 - v.leading_zeros() //~ manual_highest_one
+    } else {
+        let _ = "Rust is fast, memory-safe and productive.";
+        3
+    };
+    let _ = u32::BITS.checked_sub(u.leading_zeros() + 1);
```
Comment (CommanderStorm) [inline tests/ui/manual_highest_one.rs]: 🤔 Are the false negatives something we can get rid of?
Aliasing I can understand, that is a bit hard. But the second one?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
community review: Except Gri-ffin's suggestions, not much I could find. Seems well scoped overall.

I have a few questions about some of the false negatives, but nothing major I think

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/9117fa75789b37e502b24ac47015e778d80cdab1..ced8d1c49665d794af1a1d829ffb66bfbf028262)*

File: tests/ui/manual_highest_one.rs (hunk 12 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs:137]: I finally understand which kinds of macro usages are allowed. For example, `macro!().bit_widrh() - 1` should be linted because `macro` can be modified without side effect as long as the `bit_width` call remains valid.

File: clippy_lints/src/manual_highest_one.rs (hunk 14 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: Fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 13 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: fixed

File: clippy_lints/src/manual_highest_one.rs (hunk 16 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: done

File: tests/ui/manual_highest_one.rs (hunk 19 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs]: It is equivalent to the following:

```rust
do_something();
u.highest_one().unwrap_or_else(|| todo!());    
```

However, this should belong to `blocks_in_conditions`, not `manual_highest_one`. So, this part will be removed.

File: tests/ui/manual_highest_one.rs (hunk 12 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs:137]: To my understanding, `Span` carries no information about `identity!` macro because the macro does nothing and span is not updated during expansion. This is a corner case. IMO, we have two options:

1. Introduce `EarlyLintPass` and tackle the AST.
2. Ignore the case because it is unlikely to matter in practice. Appliciability can be also lowered.

Comment (Gri-ffin) [reply tests/ui/manual_highest_one.rs:137]: Changing this lint to an early pass would be an unecessary complexity for an edge case, it's better to just not deal with macros.

Comment (Jarcho) [reply tests/ui/manual_highest_one.rs:137]: Early passes also run after macro expansion so it wouldn't help anyway. There's currently no way to detect an identity macro (`$e:expr => $e`).

Comment (qdot3) [reply tests/ui/manual_highest_one.rs:137]: I will ignore `identity!` macro and leave it a false positive, lowering applicaibility to `MaybeIncorrect`.

File: tests/ui/manual_highest_one.rs (hunk 20 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs]: The second pattern is now supported.

File: tests/ui/manual_highest_one.rs (hunk 17 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs]: Do you mean to replace `do_something` and `may_change_state` with `black_box`? If so, it seems a resonable choice because `black_box` is descriptive enough.

File: tests/ui/manual_highest_one.rs (hunk 18 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs]: No. It is not for `manual_highest_one` and removed.

File: clippy_lints/src/manual_highest_one.rs (hunk 7 above)
Comment (qdot3) [reply clippy_lints/src/manual_highest_one.rs]: Fixed.

`manual_highest_one` and `manual_ilog2` share some patterns. For now, I simply allow them, but is there a better choice?

File: tests/ui/manual_highest_one.rs (hunk 15 above)
Comment (qdot3) [reply tests/ui/manual_highest_one.rs]: Applied.

**Review by CommanderStorm — APPROVED** [review]
(community review)
Lgtm from my side. 
Make sure to squash the changes into one commit because of rebase merging ^^

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17472/49e2f89a2605a9a340e04955fdbefa9803e6027a..1017af2cea14aaf1bfdd7cca6a30fb2a3ccaab63)*
